[EXPERIMENTAL] new mount type: introspection

AkihiroSuda opened the pull request:
**CURRENT STATUS (Apr 6, 2017)**:  discussing the design about "scope": https://github.com/docker/docker/pull/26331#issuecomment-292068427

 - - -

## What I did

Introduced a new mount type: "introspection".

The introspection mount is a new feature that allows you to introspect the
metadata about the container from the container itself, via a procfs-like
filesystem.
### Hierarchy

If you enable the introspection mount, following files are created under the mount point:
- `container/id`: ID string of the container. e.g. `2f3cc2b029e0ca46564d5a5e38772b09947056f3b22b6a114054a468382e872e\n`
- `container/name`: Name of the container. e.g. `nginx.3.8tc0va0kw59rbbdh5x3iqc3v9\n`
- `container/fullname`: Full name of the container. e.g. `/nginx.3.8tc0va0kw59rbbdh5x3iqc3v9\n`
- `container/labels/{LABELNAME}`: Label of the container. e.g. the content of `container/labels/com.docker.swarm.service.name` would be `nginx\n`
- `daemon/name`: Hostname of the daemon node. e.g. `host01\n`

For Swarm task containers running on a manager node, following files appear as well:
- `service/id`: ID string of the service. e.g. `6h7nic7tsv16cfo0qhywj7bsh\n`
- `service/name`: Name of the service .e.g. `nginx\n`
- `task/id`: ID string of the task. e.g. `8tc0va0kw59rbbdh5x3iqc3v9\n`
- `task/name`: Name of the task. e.g. `nginx.3.8tc0va0kw59rbbdh5x3iqc3v9`
- `task/slot`: Slot number (1-based index for replicas) of the task. e.g. `1\n`. Please also refer to [the documentation of the Swarmkit](https://github.com/docker/swarmkit/blob/master/design/task_model.md#slot-model). Note that there are cases where a slot may h
ave multiple tasks with the desired state of `RUNNING`.

### Use cases for the introspection mount

Below are some example use cases for the introspection mount.
#### Deploying a service that requires the task slot number (e.g. Apache ZooKeeper)

Apache ZooKeeper is a highly available coordination service that is used by
distributed systems such as Hadoop.  A typical configuration file (`zoo.cfg`)
for ZooKeeper would be as follows:

```
tickTime=2000
dataDir=/var/lib/zookeeper
clientPort=2181
initLimit=5
syncLimit=2
server.1=zoo1:2888:3888
server.2=zoo2:2888:3888
server.3=zoo3:2888:3888
```

ZooKeeper also requires a file named `myid` to be located under `dataDir`.
The content of `myid` is `1\n` for `server.1`, `2` for `server.2`, and so on.

The `task/slot` file under the introspection mount can be used for generating
the `myid` file.

See also #24110
#### 3rd party job scheduler

A 3rd party job scheduler can be built on a Docker service using the
introspection mount.

For example, the `task/slot` file under the introspection mount can be used for
implementing a scheduler that executes multiple batch jobs in parallel.
(Similar to the `{%}` symbol in the GNU parallel.)

See also #23843
#### 3rd party orchestration/monitoring tool

A container can send the `container/id` file under the introspection mount to
some 3rd party orchestration/monitoring tool.  Then such a tool take appropriate
action using the ID information.

See also #7685
## How I did it

Added a new mount type called "introspection".

The filesystem hiearchy is defined as the `RuntimeContext` struct in `daemon/rtcontext.go`.
The struct is mapped to the actual files via `daemon/introspection.go`.
## How to verify it
Start the daemon with `--experimental` flag.

Then create a service with the introspection mount `/foo`:

```
$ docker service create --name nginx --replicas 3 --mount type=introspection,introspection-scope=.,dst=/foo nginx
```

Enter a container and read the files under `/foo`:

```
$ docker exec -it $(docker ps -q -f label=com.docker.swarm.service.name=nginx | head -n 1) sh
# find /foo
/foo
/foo/container
/foo/container/name
/foo/container/labels
/foo/container/labels/com.docker.swarm.service.id
/foo/container/labels/com.docker.swarm.node.id
/foo/container/labels/com.docker.swarm.task.name
/foo/container/labels/com.docker.swarm.task
/foo/container/labels/com.docker.swarm.task.id
/foo/container/labels/com.docker.swarm.service.name
/foo/container/id
/foo/container/fullname
/foo/task
/foo/task/name
/foo/task/slot
/foo/task/id
/foo/service
/foo/service/name
/foo/service/id
/foo/daemon
/foo/daemon/name
# cat /foo/task/slot
3
```

Note that the `service` directory and the `task` directory can be seen on manager nodes in the current implementation.

## Description for the changelog

new mount type: introspection
## A picture of a cute animal (not mandatory but encouraged)

![penguin](http://www.elstonhill.com/sitebuilder/images/DSCN4660a-455x600.jpg)

Replaces #24893
Fix #24110
Update #23843, #24113, #26318

CC (those whose name appeared at #24893): 
@stevvooe @justincormack @SvenDowideit @thaJeztah @dweomer @vbatts @dmcgowan @crosbymichael @tonistiigi

The code works, but maybe merging this PR should be deferred until several relevant PRs get merged (e.g. https://github.com/docker/swarmkit/pull/1563, https://github.com/docker/docker/pull/26837, https://github.com/docker/docker/pull/26825)

TODOs
- [x] more tests
- [x] migrate to the new experimental system (https://github.com/docker/docker/pull/27223)
- [X] clarification about the use cases (note that multiple containers running simultaneously can share the single slot number, https://github.com/docker/swarmkit/pull/1741)
- [x] MS Windows
- [x] Support for `docker run`, via long-type volume option e.g. `--volume type=introspection,dst=...` (#28527)
- [X] Scope limitation https://github.com/docker/docker/pull/26331#issuecomment-292068427
- [ ] vendor https://github.com/docker/swarmkit/pull/1642


Some ideas might worth adding in future (in another PR)
- CLI alias (e.g. `--introspection`) for `--volume type=introspection,dst=$DEFAULT_DST`. We need to determine the default mount point for both UNIX and Windows.
-  `/node/labels` (dynamic)
-  `/service/labels` (dynamic)
- support `/service` and `/task` for containers running on worker nodes
- port information. e.g. `/ports/tcp/80/HostPort`. suggested by @justincormack: https://github.com/docker/docker/pull/26331#issuecomment-245242041
- memory allocation information. e.g. `/container/cgroup/memory`?. suggested by @justincormack: https://github.com/docker/docker/pull/26331#issuecomment-248265460
- machine-friendly version of `nslookup tasks.nginx`. e.g. `/service/tasks`? https://github.com/docker/docker/pull/26331#issuecomment-245842177

For reference, an older version of this PR (which was implemented as a built-in volume driver) is available at https://web.archive.org/web/20161014053843/https://github.com/docker/docker/pull/26331 .

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp

AkihiroSuda committed:
[EXPERIMENTAL] new mount type: introspection

This commit introduces a new mount type called "introspection".
The introspection mount allows users to introspect the metadata about the
container from the container itself, via a procfs-like filesystem.

How to test:

    $ docker service create --name nginx --replicas 3 --mount type=introspection,introspection-scope=.,dst=/foo nginx
    $ for f in $(docker ps -q -f label=com.docker.swarm.service.name=nginx);do docker exec $f sh -c 'for g in $(find /foo -type f); do echo ===$g===; cat $g; done'; done

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

justincormack commented:
Hi, thanks, taking a look.

I really think `map[string]string` needs to be individual files ie a file for each label. If we are going to have files you should not have to do any parsing, not microformats. Especially as labels can be multiple paragraphs and contain line breaks.

stevvooe commented:
@justincormack Is there a way we can do that without having user input become filenames?

justincormack commented:
@stevvooe hmm, probably no. We might have to filter keys that are not within the recommended guidelines for naming too.

AkihiroSuda commented:
Thank you for the comment and it makes sense :smiley:


2 questions:

1. WDYT about escaping filenames? I think we can use `net/url.QueryEscape`.



2. WDYT about dynamic label update #21721 ? There would be an atomicity problem if we allow updating multiple labels at once. 

If we support updating multiple lables at once, I suggest creating a symlink from `labels` to a `labels.snapshot.XXXXX` directory, but I feel it is not pretty. So my opinion is not to support updating multiple labels at once.

justincormack commented:
1. We already recommend strongly that label names be DNS names, so I would recommend only adding names that are valid DNS names to this interface (ie no initial `.`, `0-9a-zA-Z-.`, 64 characters max length). These should then be valid filenames on all systems, otherwise there will be other issues. Labels that are not DNS names would not be available in this interface at all.
2. I don't think we should make any guarantee that multiple label updates are in any way a "transaction", in this or any other interface. If you have dependent data, put that as the value of a single label (eg hostname plus port; complex cases the user can have a json value for atomicity). So long as each individual label update is atomic that is fine.

justincormack commented:
Ok, trying it out a bit more, re other discussion points:

1. re `/` in container name, hmm, not sure. This is confusing, as it is still not used anywhere.
2. yes definitely remove the swarm directories if not a service, then it is obvious how to see if you are a service or not, it looks odd having them for non swarm containers.
3. I don't think we need the `v1` prefix. We can deprecate fields and add new ones, if we made radical changes we could add a v2 volume driver, as with this model you still have to choose to mount it.

My other concern at present is the container can't see where its introspection mountpoint is, so if it is mounted in a place it does not expect the scripts in the container won't work. I wonder if we should have a cli option `--introspect` which is short for say `-v docker_introspection:/docker` and so have a recommended mountpoint but still let you override it if you want?

justincormack commented:
One other thing people have asked for is host port info for published ports. We could add eg `ports/tcp/80/HostPort` for each published port. There is an issue about `HostIp` which is normally `0.0.0.0` which is not very helpful. This is also different in swarm mode, as you cant guarantee to reach a particular container, so maybe there are too many complexities to this for now (lets decide about design first).

justincormack commented:
The test failures don 't like the imports:
```
10:32:16 # github.com/docker/docker/daemon/cluster/executor/container
10:32:16 import cycle not allowed in test
10:32:16 package github.com/docker/docker/daemon/cluster/executor/container (test)
10:32:16 	imports github.com/docker/docker/daemon
10:32:16 	imports github.com/docker/docker/daemon/cluster
10:32:16 	imports github.com/docker/docker/daemon/cluster/executor/container
10:32:16 
10:32:16 FAIL	github.com/docker/docker/daemon/cluster/executor/container [setup failed]
```

justincormack commented:
Also needs a rebase...

justincormack commented:
@AkihiroSuda I think generally I like the overall design now. There is a bunch of stuff to resolve still and some decisions to make, but I think it is the right general direction...

AkihiroSuda commented:
@justincormack Thank you, I'll update this PR ASAP

AkihiroSuda commented:
@justincormack Updated PR, please look into this?

## Changes

### Filesystem View

For https://github.com/docker/docker/pull/26331#issuecomment-245105201, https://github.com/docker/docker/pull/26331#issuecomment-245229962

* Now `map[string]string` are separate files.  (`daemon/introspect.go`)
* Prohibit a label that contains `/`, `\`, or `:` from appearing at the filesystem. For keeping compatibility, I didn't make the daemon itself to reject these bad labels. The introspector skips such a label and just calls `logrus.Warn`.  (`daemon/introspect.go`)

For https://github.com/docker/docker/pull/26331#issuecomment-245239605

* Removed '/' prefix from the value of `/container/name`. However, I introduced `/container/fullname` that has the '/' prefix. I'm not sure we need this `container/fullname` file. (`daemon/rtcontext.go`)
* Hide `/task`, `/service`, and `node` if the container is not a swarm task. (`daemon/rtcontext.go`, `daemon/introspect.go`)
* Removed v1 prefix. We could safely add a v2 driver for future radical changes. (`daemon/rtcontext.go`)

Others (RFC!):

 * `/task/name` was always empty. So I switched to use the value of `Container.Labels["com.docker.swarm.task.name"]` rather than `types/swarm.Task.Name`.
 * There has been already dynamic label update for swarm related things. I removed these `label` files just for ease of the initial implementation.
 * Introduced `/daemon` (corresponds to `types.Info`). Currently, only `/daemon/name` and `/daemon/labels` are exposed. I think `/daemon/labels` is useful but not sure about `/daemon/name`.


### Build

 * Rebased to upstream
 * Avoid import cyle (https://github.com/docker/docker/pull/26331#issuecomment-245242355) by removing a import from `github.com/docker/docker/daemon/cluster` to `github.com/docker/docker/daemon`, and introducing the `IntrospectableCluster` interface in the `daemon` package. (`daemon/cluster.go`, `daemon/daemon.go`)

## Future work (IMO they should be separate PRs?)

 * Shorter CLI. e.g. (`--introspect` as an alias for `-v docker_introspection:/docker`). We need to determine the default path. https://github.com/docker/docker/pull/26331#issuecomment-245239605
 * Revive swarm-related label introspection
 * Expose network related information (e.g. `ports/tcp/80/HostPort`. My colleague also recently asked for this :smiley:) We need to determine the fields and its hierarchy. https://github.com/docker/docker/pull/26331#issuecomment-245242041
 * Provide task discovery? (machine-friendly version of `nslookup tasks.nginx`)

----

DEMO

```console
$ docker run --rm  -v docker_introspection:/foo busybox sh -c 'for f in $(find /foo -type f);do echo ===$f===; cat $f; done'
===/foo/container/name===
romantic_noether
===/foo/container/id===
3b5f9e1bacf85424dbbbbd1373a0ae17ae00504fa1618059fa85bf9f4ad29e30
===/foo/container/fullname===
/romantic_noether
===/foo/daemon/name===
ws01
```

```console
$ docker service create --name nginx --replicas 3 --mount type=volume,src=docker_introspection,dst=/foo nginx
$ docker exec $SOME_NGINX_CONTAINER bash -c 'for f in $(find /foo -type f);do echo ===$f===; cat $f; done'
===/foo/container/name===
nginx.1.8toszsdz7duypbbl8uwehlrbs
===/foo/container/labels/com.docker.swarm.service.id===
0g8upo0xzr0l7pnw41jcdpt1d
===/foo/container/labels/com.docker.swarm.node.id===
clhlkxk2gzbk3edhropygbknq
===/foo/container/labels/com.docker.swarm.task.name===
nginx.1
===/foo/container/labels/com.docker.swarm.task===
===/foo/container/labels/com.docker.swarm.task.id===
8toszsdz7duypbbl8uwehlrbs
===/foo/container/labels/com.docker.swarm.service.name===
nginx
===/foo/container/id===
e7a6c58f3cbfd6754158d8db284e08396df0cadafa1a4c0e338826a012d9ef67
===/foo/container/fullname===
/nginx.1.8toszsdz7duypbbl8uwehlrbs
===/foo/task/name===
nginx.1
===/foo/task/slot===
1
===/foo/task/id===
8toszsdz7duypbbl8uwehlrbs
===/foo/service/name===
nginx
===/foo/service/id===
0g8upo0xzr0l7pnw41jcdpt1d
===/foo/daemon/name===
ws01

```

Now I'm going to add tests and docs

justincormack commented:
The tests need tweaking to special case this volume driver:
```
08:16:06 
08:16:06 ----------------------------------------------------------------------
08:16:06 FAIL: docker_cli_run_test.go:4301: DockerSuite.TestRunNamedVolumeNotRemoved
08:16:06 
08:16:06 docker_cli_run_test.go:4309:
08:16:06     c.Assert(strings.TrimSpace(out), checker.Equals, "test")
08:16:06 ... obtained string = "" +
08:16:06 ...     "docker_introspection\n" +
08:16:06 ...     "test"
08:16:06 ... expected string = "test"
08:16:06 
08:16:07 
08:16:07 ----------------------------------------------------------------------
08:16:07 FAIL: docker_cli_run_test.go:4318: DockerSuite.TestRunNamedVolumesFromNotRemoved
08:16:07 
08:16:07 docker_cli_run_test.go:4331:
08:16:07     c.Assert(strings.TrimSpace(out), checker.Equals, "test")
08:16:07 ... obtained string = "" +
08:16:07 ...     "docker_introspection\n" +
08:16:07 ...     "test"
08:16:07 ... expected string = "test"
08:16:07 
08:16:07 
```

AkihiroSuda commented:
Now CI is almost passing except a flaky test https://github.com/docker/docker/issues/26506 .

A basic integration test is added as `DockerSwarmSuite.TestSwarmServiceWithIntrospectionVolume`.
I'll add more assertions (after https://github.com/docker/docker/pull/26519 gets merged, for ease of integration CLI testing)

justincormack commented:
@AkihiroSuda one thing we discussed, and is (almost) possible now that https://github.com/docker/docker/pull/22373 is merged is that rather than using a volume driver we could use a special mount type (this would also require CLI support for swarm style mounts, but I think this is planned now). ie `docker run --mount type=introspection,target=/run/introspection ...`

Still needs some thought though cc @cpuguy83 @ehazlett

justincormack commented:
There is an issue here https://github.com/docker/docker/issues/8427 that wants to know memory allocation. Just linking - I don't think we should necessarily try to get everything into the initial PR.

justincormack commented:
Needs a rebase...

justincormack commented:
Design LGTM. We need to decide on the `--mount` issue, I prefer `--mount` so this is consistent, and you do not see the introspection volume driver listed, it seems cleaner. This still requires this to be added for normal containers, so can be done later.

justincormack commented:
cc @icecrime

AkihiroSuda commented:
Sorry for being late😭 , I'll update ASAP

Yes, I prefer --mount as well.
Also, the memory allocation problem seems very interesting, thank you for sharing!

icecrime commented:
First thing first: that's amazing @AkihiroSuda :astonished: Thanks a lot!

- I really like that this solves a ton of user problems with what is a very reasonable amount of code.
- It's a bit scary that we're introducing a whole new API, but maybe that's the way to go. We know for sure that @shykes is a bit advocate for filesystem interface, so I'm sure he'll love the proposal. I will try to bring it to his attention as it's quite a significant change.
- If we merge this, we should probably document what scope we want to give to it, because I can picture this going in a lot of different directions very fast (adding everything `docker info` has, adding mutating operations, etc, etc).
- I think we should give this a try and ship it in experimental at first to gather some feedback.

Design LGTM :+1:

AkihiroSuda commented:
@icecrime I'm glad you like it :smiley: 
I'll migrate the introspection fs to the new mount type, when https://github.com/docker/swarmkit/pull/1563 gets merged

AkihiroSuda commented:
Update PR.

Usage: `docker service create --name nginx --replicas 3 --mount type=introspection,dst=/foo nginx`

The result should be identical to https://github.com/docker/docker/pull/26331#issuecomment-245842177

Changes and TODOs (must be in this PR):
- [X] built-in volume --> new mount type
- [X] Add `experimental` build tag
- [ ] Rebase, when several relevant PRs get merged (https://github.com/docker/swarmkit/pull/1563, https://github.com/docker/docker/pull/26837, https://github.com/docker/docker/pull/26825)
- [ ] Document
- [ ] More integration test cases

TODOs (in future separate PRs, IMO)
- [ ] MS Windows
- [ ] Alias for `--mount type=introspection,dst=$DEFAULT_DST`. We need to determine the default mount point for both UNIX and Windows.
- [ ] `/node/labels` (dynamic)
- [ ] `/service/labels` (dynamic)

TODOs? (should be in future separate PRs)
- [ ] port information. e.g. `/ports/tcp/80/HostPort`. suggested by @justincormack: https://github.com/docker/docker/pull/26331#issuecomment-245242041
- [ ] memory allocation information. e.g. `/container/cgroup/memory`?. suggested by @justincormack: https://github.com/docker/docker/pull/26331#issuecomment-248265460
- [ ] machine-friendly version of `nslookup tasks.nginx`. e.g. `/service/tasks`? https://github.com/docker/docker/pull/26331#issuecomment-245842177

AkihiroSuda commented:
CI failure is legal but will be fixed when the equivalent of https://github.com/docker/swarmkit/pull/1563 is merged to the engine

```
00:50:16.996 ----------------------------------------------------------------------
00:50:16.996 FAIL: docker_cli_service_create_test.go:15: DockerSwarmSuite.TestServiceCreateMountVolume
00:50:16.997 
00:50:16.997 [d5a0135aa6841] waiting for daemon to start
00:50:16.997 [d5a0135aa6841] daemon started
00:50:16.997 docker_cli_service_create_test.go:33:
00:50:16.997     waitAndAssert(c, defaultReconciliationTimeout, func(c *check.C) (interface{}, check.CommentInterface) {
00:50:16.997         if task.NodeID == "" || task.Status.ContainerStatus.ContainerID == "" {
00:50:16.998             task = d.getTask(c, task.ID)
00:50:16.998         }
00:50:16.998         return task.NodeID != "" && task.Status.ContainerStatus.ContainerID != "", nil
00:50:16.998     }, checker.Equals, true)
00:50:16.998 daemon_swarm.go:200:
00:50:16.998     c.Assert(status, checker.Equals, http.StatusOK, check.Commentf("output: %q", string(out)))
00:50:16.998 ... obtained int = 404
00:50:16.999 ... expected int = 200
00:50:16.999 ... output: "{\"message\":\"task 9yxtuq9zv4xqjadxk8nl4j9z2 not found\"}\n"
00:50:16.999 
00:50:16.999 [d5a0135aa6841] exiting daemon
00:50:22.678 
00:50:22.678 ----------------------------------------------------------------------
```

(caused by error `Duplicate mount point '/foo'` from https://github.com/docker/docker/pull/26331/files#diff-9651468cafd3eee09570f3733efd7dc4R381 , will be fixed in the equivalent of https://github.com/docker/swarmkit/pull/1563/files#diff-b7d35a34ddaff6d0a7546923c4f67111L92)

AkihiroSuda commented:
Added a documentation (`experimental/introspection.md`) and updated the description text of this PR (thank you for your suggestion at Berlin @tiborvass )

This contains the "use cases" section.

AkihiroSuda commented on daemon/cluster/executor/container/container.go:
```diff
@@ -124,7 +125,6 @@ func (c *containerConfig) config() *enginecontainer.Config {
 		Env:        c.spec().Env,
 		WorkingDir: c.spec().Dir,
 		Image:      c.image(),
-		Volumes:    c.volumes(),
```
Changes in this `daemon/cluster/executor/container/container.go` is from https://github.com/docker/swarmkit/pull/1563

stevvooe commented:
@AkihiroSuda Take a peak at https://github.com/docker/swarmkit/pull/1650 for possibly hitting this use case. It allows you to inject contextual parameter into env and mount declarations.

AkihiroSuda commented:
@stevvooe Thank you for the information. Yes, it SGTM and I agree it can hit the use cases as well.
I think it is ok to have the both features.

Compared to the template, this PR still has some advantages:
- Can support dynamic update
- Can support non-swarm information (e.g. `container/id`, `container/labels`, and @justincormack 's suggestion about memory https://github.com/docker/docker/pull/26331#issuecomment-248265460)

The advantages of the template would be:
- Can limit the scope of the information exposed to containers
- Can expand mount spec

LK4D4 commented:
@AkihiroSuda let's rebase it.

AkihiroSuda commented:
@LK4D4 done

thaJeztah commented:
@AkihiroSuda vendor is failing;

```
10:17:35 The result of vndr differs
10:17:35 
10:17:35  M vendor/github.com/docker/swarmkit/api/types.pb.go
10:17:35  M vendor/github.com/docker/swarmkit/api/types.proto
10:17:35 
10:17:35 Please vendor your package with github.com/LK4D4/vndr.
```

And looks like one test is failing is well;

```
11:10:41 ----------------------------------------------------------------------
11:10:41 FAIL: docker_cli_swarm_experimental_test.go:84: DockerSwarmSuite.TestSwarmServiceWithIntrospectionVolume
11:10:41 
11:10:41 [d56f8ca0d5c3d] waiting for daemon to start
11:10:41 [d56f8ca0d5c3d] daemon started
11:10:41 docker_cli_swarm_experimental_test.go:97:
11:10:41     // make sure task has been deployed.
11:10:41     waitAndAssert(c, defaultReconciliationTimeout, d.checkActiveContainerCount, checker.Equals, replicas)
11:10:41 docker_utils.go:1571:
11:10:41     c.Assert(v, checker, args...)
11:10:41 ... obtained int = 0
11:10:41 ... expected int = 3
11:10:41 
11:10:41 [d56f8ca0d5c3d] exiting daemon
```

AkihiroSuda commented:
@thaJeztah
Thank you for pointing out the test failure, it should be fixed now 😅 

For vendoring Swarmkit, we need to wait for https://github.com/docker/swarmkit/pull/1642

thaJeztah commented:
Oh, silly me; there's still a "needs-vendoring" label :smile: ignore ignore

AkihiroSuda commented:
added tests (`docker_cli_run_experimental_test.go`)

AkihiroSuda commented:
updating doc as well

```diff
diff --git a/experimental/introspection.md b/experimental/introspection.md
index 343f03b..cda1e44 100644
--- a/experimental/introspection.md
+++ b/experimental/introspection.md
@@ -15,14 +15,14 @@ If you enable the introspection mount, following files are created under the mou
 
 - `daemon/name`: Hostname of the daemon node. e.g. `host01\n`
 
-For Swarm task containers, following files appear as well:
+For Swarm task containers running on a manager node, following files appear as well:
 
 - `service/id`: ID string of the service. e.g. `6h7nic7tsv16cfo0qhywj7bsh\n`
 - `service/name`: Name of the service .e.g. `nginx\n`
 
 - `task/id`: ID string of the task. e.g. `8tc0va0kw59rbbdh5x3iqc3v9\n`
 - `task/name`: Name of the task. e.g. `nginx.3.8tc0va0kw59rbbdh5x3iqc3v9`
-- `task/slot`: Slot number (1-based index for replicas) of the task. e.g. `1\n`. Please also refer to [the documentation of the Swa
rmkit](https://github.com/docker/swarmkit/blob/master/design/task_model.md#slot-model).
+- `task/slot`: Slot number (1-based index for replicas) of the task. e.g. `1\n`. Please also refer to [the documentation of the Swa
rmkit](https://github.com/docker/swarmkit/blob/master/design/task_model.md#slot-model). Note that there are cases where a slot may h
ave multiple tasks with the desired state of `RUNNING`.
 
 ## Use cases for the introspection mount
 

```

yank1 commented:
LGTM

jhowardmsft commented:
@thaJeztah Tried it on Windows, but the rebase was a little painful (to get swarm support on Windows), and to make it build at all on Windows (the build files aren't in the repo) @AkihiroSuda if you can rebase, I can figure out how to support Windows. My attempt is at https://github.com/Microsoft/docker/commit/0b6bf3422871ef82c7d5a4b71f07f59c3b4dae05 - it should only be the volume_windows.go bit that's needed if rebased on master latest. It could be tidied too as there's now a bunch of common code which should be duplicated.

AkihiroSuda commented:
@jhowardmsft 

Sorry for the trouble, I'll rebase ASAP this week (or next week maybe).

Note that the introspection support for `docker run` is broken because we removed `docker run --mount type=(bind|volume|tmpfs|introspection)` (introduced in https://github.com/docker/docker/pull/26825 , removed in https://github.com/docker/docker/pull/28838).


I'll also try to implement the long-syntax `--volume` (replaces `--mount`) , so as to revive the introspection support for `docker run` .

https://github.com/docker/docker/issues/28527#issuecomment-262817805

> deprecate --mount (maybe remove it from run immediately) and have --volume support the long syntax.

(cc @icecrime)

AkihiroSuda commented:
Rebased to master and confirmed it works on a local Linux host.
Hope CI green :pray:

AkihiroSuda commented:
@jhowardmsft 
I confirmed the introspection seems working on Windows as well :smile: . PTAL.

```
Microsoft Windows [Version 10.0.14393]
(c) 2016 Microsoft Corporation. All rights reserved.

C:\>dir
 Volume in drive C has no label.
 Volume Serial Number is 1CA2-5DC7

 Directory of C:\

01/31/2017  06:55 PM    <SYMLINKD>     introspection [\\?\ContainerMappedDirectories\3F8F6A34-94F6-48FB-8F59-7788F0EC4BF
F]
11/20/2016  08:32 PM             1,894 License.txt
01/31/2017  06:57 PM    <DIR>          Program Files
07/16/2016  09:09 PM    <DIR>          Program Files (x86)
01/31/2017  06:55 PM    <DIR>          Users
01/31/2017  06:56 PM    <DIR>          Windows
               1 File(s)          1,894 bytes
               5 Dir(s)  21,222,645,760 bytes free

C:\>dir introspection
 Volume in drive C has no label.
 Volume Serial Number is 1CA2-5DC7

 Directory of C:\introspection

01/31/2017  06:55 PM    <DIR>          .
01/31/2017  06:55 PM    <DIR>          ..
01/31/2017  06:56 PM    <DIR>          container
01/31/2017  06:56 PM    <DIR>          daemon
               0 File(s)              0 bytes
               4 Dir(s)  87,792,730,112 bytes free

C:\>dir introspection\container
 Volume in drive C has no label.
 Volume Serial Number is 1CA2-5DC7

 Directory of C:\introspection\container

01/31/2017  06:56 PM    <DIR>          .
01/31/2017  06:56 PM    <DIR>          ..
01/31/2017  06:56 PM                17 fullname
01/31/2017  06:56 PM                65 id
01/31/2017  06:56 PM                16 name
               3 File(s)             98 bytes
               2 Dir(s)  87,792,730,112 bytes free

C:\>type introspection\container\name
sad_nightingale

C:\>dir introspection\daemon
 Volume in drive C has no label.
 Volume Serial Number is 1CA2-5DC7

 Directory of C:\introspection\daemon

01/31/2017  06:56 PM    <DIR>          .
01/31/2017  06:56 PM    <DIR>          ..
01/31/2017  06:56 PM                 6 name
               1 File(s)              6 bytes
               2 Dir(s)  87,792,730,112 bytes free

C:\>type introspection\daemon\name
win01
```

For Linux, we could use `docker service create --mount type=introspection` for testing this feature.
However, since `docker service` does not support Windows, I added a small util (`contrib/tmp-introspection-windows/main.go`) for creating a non-service container with `HostConfig.Mounts[0].Type = TypeIntrospection`.

When we reached a conclusion about `docker run --mount` (28527), we can remove  `contrib/tmp-introspection-windows/main.go`.

stevvooe commented:
After the removal of `--mount`, I am not sure how we can support this. Let's add mounts back in so we can move forward.

cirocosta commented:
@stevvooe could you link the relevant PRs/issues regarding `--mount` so we can keep track here? thx!

stevvooe commented:
@cirocosta It was removed in https://github.com/docker/docker/pull/28838. Not sure why. It was done hastily over Thanksgiving.

krasi-georgiev commented:
just to add my use case - improving the percona mysql cluster scaling and every container need to know all other running percona containers so it connects and syncs
the entrypoint on each starting container runs 
$joinNodes=dig tasks.$serviceName 

at the moment serviceName is set as -e serviceName=... in the service create command
but after this PR it would simply be
$joinNodes=dig tasks.$(cat /service/name)

so this will need the supported on all  swarm nodes to make it work.

krasi-georgiev commented:
some thoughts

[UX] - this has so much potential that should be enabled by default without any need to use the mount option

technically would it be much difficult to use the dns query approach to - dig tasks.service

otherwise we would have 2 roadmaps for the same thing - 
   querying services - dns based
   querying container - file based

I think it should be one or the other.

AkihiroSuda commented:
@krasi-georgiev

Thank you for feedback.

> so this will need the supported on all swarm nodes to make it work.

Currently some swarm-specific contents are unvisible on worker nodes because of the limitation of swarmkit (as you cannot run `docker service ..` and `docker node ..` on worker nodes).

However, you can still access the `labels` directory even on worker nodes:
`cat /foo/container/labels/com.docker.swarm.service.name`

> [UX] - this has so much potential that should be enabled by default without any need to use the mount option

I'm considering introducing `--introspection` as an alias for `--mount type=introspection,dst=/some/default/location`.


>technically would it be much difficult to use the dns query approach to - dig tasks.service
>
>otherwise we would have 2 roadmaps for the same thing -
>querying services - dns based
>querying container - file based
>
>shouldn't be one or the other?

IMO it is ok to have both

panga commented:
I suspect we will face file naming issues with invalid characters and long paths restrictions on Windows.

What about creating a JSON file (validated against a JSON schema) with all the required data?

Everyone is already used to docker inspect API (JSON) and it would be easy to add new keys and struct in the future, while being simpler than reading files.

AkihiroSuda commented:
@panga 

We had once considered using JSON, but found that separate files seem better than JSON.
https://github.com/docker/docker/issues/24110#issuecomment-234358152

IIUC, naming issues are unlikely to happen, unless there is no label that violates the guideline https://docs.docker.com/engine/userguide/labels-custom-metadata/#key-format-recommendations

panga commented:
@AkihiroSuda Sure, I'm just raising a concern here.

Creating a file system structure adds an extra complexity from my point of view.
Nowadays everybody is used to JSON and simpler approaches such as JSON file or HTTP JSON, which is how AWS and OpenStack handles VM instances metadata.

'jq' and other JSON parsers are commodity tools, far better than find/cat commands to get some values.

SvenDowideit commented:
is `/sys/docker` or `/sys/runc` a good default location? (I like the idea of `--introspection`...

krasi-georgiev commented:
@SvenDowideit
> is /sys/docker or /sys/runc a good default location?

I think the default should be 
`/run/.....`

to follow the defaults used for secrets
`/run/secrets/...`

krasi-georgiev commented:
> However, you can still access the labels directory even on worker nodes:
> cat /foo/container/labels/com.docker.swarm.service.name

@AkihiroSuda from a quick peek on the code changes it looks like you still need to set com.docker.swarm.service.name using `-f label=com.docker.swarm.service.name=nginx` right ? 
If that is true I  think doesn't make it more convenient that just using env variables.

Is there is no way to auto populate com.docker.swarm.service.name?

this is my use case : https://github.com/docker/docker/pull/26331#issuecomment-277270441

AkihiroSuda commented:
@panga

> Creating a file system structure adds an extra complexity from my point of view.
Nowadays everybody is used to JSON and simpler approaches such as JSON file or HTTP JSON, which is how AWS and OpenStack handles VM instances metadata.

I think we need to keep separate files for UX, but I think it is ok to add JSON as well for atomicity. (in future another PR)
cc @stevvooe WDYT?

@SvenDowideit @krasi-georgiev

> is /sys/docker or /sys/runc a good default location?

I'm not sure we can make a mountpoint under sysfs.
So I'm +1 for `/run/...`.
I think the default location can be discussed after this PR gets merged.

@krasi-georgiev

> from a quick peek on the code changes it looks like you still need to set com.docker.swarm.service.name using -f label=com.docker.swarm.service.name=nginx right ?

No, it is automatically set (even on worker nodes).
https://github.com/docker/swarmkit/blob/4dbc0a367de19b9f0dfb64302cd9fd28b5909b66/agent/exec/dockerapi/container.go#L226

thaJeztah commented:
Giving this another look, and I still like this proposal; some thoughts;

- I agree on having a default location for this (`/run/<something>` sounds like a good choice)
- This needs a way to _scope_ / _limit_ what parts are exposed, especially if more information is added. There may be cases where you only want _certain_ information available to the containers, but hide other information. Perhaps `--scope=/container/labels --scope=/task` to only add container-labels, and task information, but omit other parts
- information about the node (labels? ip-address? etc), and ports that are published (especially when using random ports) is mentioned before, but looks like an important part to me as a follow up.

AkihiroSuda commented:
>  Perhaps scope=/container/labels,/task to only add container-labels, and task information, but omit other parts

Thank you for the suggestion.
I wonder we can improve this syntax to be more consistent with other "LONG" mount opts.
How about: `scope-drop=*,scope-add=/container/labels,scope-add=/task/*`
(cc @icecrime @stevvooe )

AkihiroSuda commented:
I think we will be able to move this forward soon, right after https://github.com/docker/docker/pull/32251 gets merged.

@thaJeztah 
PTAL about scope-drop/add?
https://github.com/docker/docker/pull/26331#issuecomment-287964609

melo commented:
This is great news @AkihiroSuda...

Do you know if a PR for /port/* exists already? Speaking for myself, I've been waiting for a clean way to have access to this information for quite some time, and I appreciate the work put into this introspection mount as a path to reach that point, but I would like to keep track of the /port spec and progress, and push it forward in any way I can.

Thanks,

thaJeztah commented:
Oh, apologies, I noticed yesterday you had replied, but I missed that earlier.

> `scope-drop=*,scope-add=/container/labels,scope-add=/task/*`

I like that thought. Not sure if asterisks (`*`) should be used; I remember they could be troublesome with the shell possibly expanding them (but perhaps that's not the case here); alternatively;

- `scope-drop=all` (or `scope-drop=/`)
- `scope-add=/task/` is implicitly "everything under `/task/`"

For `service create` we don't use the `-add` / `-rm` suffices, hence my original "`scope=...`" which implicitly meant "drop all, but add what is specified", but I can see dropping just one part would be very verbose there, so yes, I like `scope-drop` / `scope-add`.

We'll be adding this as "experimental" feature, so we can still tweak where needed, based on feedback 👍

cirocosta commented:
👍 on replacing `*`  by `all`; it's easy to forget to use single quotes in some shells to not have the expansion

AkihiroSuda commented:
Sorry for going back and forth, but I think [the flag needs to have `introspection-` prefix](https://github.com/docker/docker/blob/master/opts/mount.go) e.g.
`--mount type=introspection,target=/run/introspection,introspection-scope-drop=/,introspection-scope-add=/container,introspection-scope-add=/daemon`...

This UX seems unacceptable :sob:

I think it would be great if we can use bash `extglob` here.
e.g.
```
--mount type=introspection,target=/run/introspection,introspection-scope=/+(container,daemon)
```

But I don't think there is any extglob implementation for Go.
Seems we need more bikeshedding :bike:


- - -

@melo
> Do you know if a PR for /port/* exists already? S

It is under discussion, but I think it should be another PR.
I think we should focus on the infra part on this PR.
https://github.com/docker/docker/pull/26331#issuecomment-245242041

thaJeztah commented:
> but I think the flag needs to have introspection- prefix 

hm, yes, that's a good point. Looking at what we do for those, I'm starting to wonder if that's very useful. Given that a `--mount` is basically using a `--type`, what advantages do the prefixes bring us?

From a UX perspective;

```
--mount type=volume,volume-label=foo=bar,volume-nocopy=true
```

is not adding much over;

```
--mount type=volume,label=foo=bar,nocopy=true
```

(imo actually _more_ confusing)

From a _code_ perspective, it may be easy to have the prefix around (if "prefix" doesn't match "type", it must be invalid), but simply splitting the validation to different methods, based on `type` would probably work equally well.

### Getting this moving

Ok, it's clear that this _will_ need feedback, but I think the best feedback we'll get is if people get their hands on it 😇. It needs to be clearly communicated as "experimental", and may need a couple of releases to evaluate UX, and features.

I'd pretty much like to see this feature moving, because it's (imo) a very useful feature. Let me try to list some options;

**_A. should the options have a prefix (`introspection-foo`)?_**

For consistency, **yes**. However, I'll open a separate issue to discuss this "in general", and see if we want to stick with having these prefixes.

**_B. should we use `-drop` / `-add`?_**

Brainstorming once more on this one, possible options;

1. by default, _drop_ all, and make `introspection-scope` **required** (could be considered "secure by default")
2. by default, include _all_, but when setting `introspection-scope=/foo`, implicitly _drop_ everything, except for `/foo` ("what's specified in `introspection-scope`)
3. by default, include _all_, but have `introspection-scope-add` / `introspection-scope-drop` to add/drop scopes

> I think it would be great if we can use bash extglob here.

If we go for `1.`, or `2.`, this could be a future extension to the syntax for those. Alternatively, have a `introspection-scope-match` option that is added; we'll have to decide if that can be used in _combination_ with the other options.

We also need to look at how that option would work for Windows containers (same syntax? back- or forward-slash?).

w.r.t. Windows/Linux; do any of the paths/scopes contain periods (`.`)? If not, perhaps using a period for scope definition would be an option, as it's a bit similar to the way templates work with `--format`;

```
scope=container.labels,tasks,container.ports

or

scope-add=container.labels,tasks,container.ports

sure, globbing would be possible, but of course shell expansion would be a thing still?

scope=container.*,foo.*.something
```

@AkihiroSuda wdyt?

thaJeztah commented:
Discussing this in the maintainers meeting, and we seem to agree that option 1 ("explicit") is the way to go; make "scope" a required option

Also discussed `/` vs `.`, and we'd prefer using a `.`

AkihiroSuda commented:
Implemented scope.

Examples:
- `--mount type=introspection,target=/foo,introspection-scope=.`
- `--mount type=introspection,target=/foo,introspection-scope=.container.labels,introspection-scope=.task`


Error if no scope is set:
```console
$ docker run -it --rm --mount type=introspection,target=/foo busybox
invalid argument "type=introspection,target=/foo" for --mount: at least one 'introspection-scope' is required. e.g. '.' (denotes all), '.containers.labels', '.task'
See 'docker run --help'.
```

Error if invalid scope is set:
```console
$ docker run -it --rm --mount type=introspection,target=/foo,introspection-scope=foo busybox
docker: Error response from daemon: invalid scope: foo (valid scopes: [. .container .container.id .container.name .container.fullname .container.labels .daemon .daemon.name .daemon.labels]).
```

A scope name needs start with a period `.`.
The scope name `.` allows all the files to be present.
This may look weird, but consistent with `--format {{json .}}` and so on.

titpetric commented:
It was mentioned higher up that DNS records would be a valid introspection API. I assume this would be done with TEXT records on individual hosts entries, and would completely drop the need to add `--mount` to any containers (and I agree, UX for introspection should be available to all without an additional mount opt). Was some work done along these lines yet?

cc: @krasi-georgiev @AkihiroSuda

AkihiroSuda commented:
@titpetric 

Not yet.
This PR just adds infrastructure part.
DNS and default mount should be separate GitHub issues/PRs

AkihiroSuda commented:
@thaJeztah PTAL?

AkihiroSuda commented:
@thaJeztah PTAL?
If SGTY, I'll split CLI part from this PR.

AkihiroSuda commented:
kindly ping :smiley: @thaJeztah

cpuguy83 commented:
What's the status here? Hard to follow along. This is labeled with code-review but some of the comments would suggest otherwise.

AkihiroSuda commented:
@cpuguy83

@thaJeztah suggested implementing "scope" before merging this PR
https://github.com/moby/moby/pull/26331#issuecomment-292120597
https://github.com/moby/moby/pull/26331#issuecomment-293835081

I implemented that feature in https://github.com/moby/moby/pull/26331#issuecomment-293835081 and waiting for his feedback :sweat_smile:

* `--mount type=introspection,target=/foo,introspection-scope=.` (enables all introspections)
* `--mount type=introspection,target=/foo,introspection-scope=.container.labels,introspection-scope=.task` (enables `/container/labels/*` and `/task/*`)

If I can get SGTM, I'll split CLI part from this PR.

jasonmp85 commented:
Is there an estimate for when it will be simple to determine a container's ID from within the container?

AkihiroSuda commented:
It is implemented in this PR but still waiting for review 😅

cpuguy83 commented:
Ping

@AkihiroSuda I like this.
Can we make scope (maybe renamed to "format") just a normal go-template on `types.ContainerJSON` with the default being the full json-ified object?

AkihiroSuda commented:
> Can we make scope (maybe renamed to "format") just a normal go-template on types.ContainerJSON with the default being the full json-ified object?

SGTM
@justincormack @thaJeztah WDYT?

vdemeester commented:
@AkihiroSuda SGTM :angel:

thaJeztah commented:
Trying to get my head round the Go template;

- First of all; this means we are good with giving access to _all_ properties of a container :+1::-1:
  - Do we want to limit what can be accessed?
  - Do we want to "abstract" it (perhaps a custom "introspection context" / struct), also to make the format consistent, even if the container JSON format may change?
- Bit in doubt over changing to `format`; IIUC we'd only use the Go template to select a _key_, or do we see uses for actually doing _formatting_ in other ways (again, trying to get my head around this 😅 - bear with me). If all we accept is the `path` to the information, `scope` could still make sense (🚲 🏠)
- What would it look like to select multiple scopes? `format={{.Foo}},{{.Bar}}`? `format={{.Foo}},format={{.Bar}}` ?
- What filename convention should we use for array-values, or are those values put in a single file? Thinking of `{{.HostConfig.Dns}}`:

    (using `docker container inspect` for reference; the introspection would look something like `--mount type=introspection,target=/foo,introspection-format='{{.HostConfig.DNS}}'`)

    ```bash
    docker container create --dns=1.1.1.1 --dns=2.2.2.2 --name foobar busybox
    docker container inspect --format '{{.HostConfig.Dns}}' foobar
    [1.1.1.1 2.2.2.2]

    ```

    Will this produce (inside the container);

    - 2 files: `/foo/HostConfig/Dns/0` and `/foo/HostConfig/Dns/1`
    - 1 file: `/foo/HostConfig/Dns`, containing `[1.1.1.1 2.2.2.2]`
    - 1 file: `/foo/HostConfig/Dns.json`, containing `["1.1.1.1","2.2.2.2"]` (or should this be optional by using `{{json .HostConfig.Dns}}`?)
- Slightly related to the above; will `{{.NetworkSettings.Networks}}`
  - create files for each network (`/foo/NetworkSettings/Networks/bridge`, `/foo/NetworkSettings/Networks/mynetwork`), each containing a (JSON?) representation of the network?
  - create directories for each network, and files for each property? (`/foo/NetworkSettings/Networks/bridge/NetworkID`, `/foo/NetworkSettings/Networks/bridge/IPAddress`)
  - Or will we require users to use a `range ..` (`{{range .NetworkSettings.Networks}}{{println .IPAddress}}{{end}}`)?
- Do we want files to be cased the same as the JSON, or (e.g.) `snake_case` or `dash-names` ?

cpuguy83 commented:
@thaJeztah My thought was to have the actual object created by the template in the mount target (which must be specified for the mount).
So if your format is `{{.HostConfig.Dns}}`, then that's the value you get in the target path rather than trying to create a tree.

AkihiroSuda commented:
As Kubernetes has got dominance now, I'd suggest using `DownwardAPIVolumeFile` with Kubernetes: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#downwardapivolumefile-v1-core

> Field | Description
> --|--
> `fieldRef`  `ObjectFieldSelector`  | Required: Selects a field of the pod: only `annotations`, `labels`, `name` and `namespace` are supported.
> `resourceFieldRef` `ResourceFieldSelector` | Selects a resource of the container: only resources limits and requests (`limits.cpu`, `limits.memory`, `requests.cpu` and `requests.memory`) are currently supported.

However, 
- `DownwardAPIVolumeFile` seems to lack some fields discussed above, e.g. port mapping and DNS. We might want to open a PR in kubernetes repo.
- If we want to bring this feature for Swarm-mode, probably we might want to hornor the design of `DownwardAPIVolumeFile` for interoperatibility. Also, maybe we want to discuss the possibility of common spec with other orchestrator (and runtime) projects.


I'm going to close this PR (Thank you all for spending your time on this PR!), but if somebody else wants to get this PR merged, please feel free to carry or just ping me.
